### PR TITLE
Allow adding labels to containers

### PR DESCRIPTION
**Pull request by carlpett**
Hi!
This PR adds the ability to set labels on the containers started by Helios. There is probably a bunch of work remaining, I haven't been able to successfully start the integration tests on my machine yet, and the building and starting of helios-solo doesn't seem to work on Windows with boot2docker? (I can understand if this is not a configuration that you test...)

But while I get that ironed out, it'd be great if you have any feedback on places I have missed to add this, or more tests that need to be written?

### CODE REVIEW

**Comment by mattnworb** [comment]
Hey @carlpett , thanks for the PR! Can you describe a bit more about the motivation behind this change?

**Comment by carlpett** [comment]
Hi!
We want to be able to set labels on the containers we deploy, and we also want to be able to set them after building the container (ie, not with a `LABEL` statement in the Dockerfile). We use this as metadata in our monitoring (cadvisor and similar expose the labels along with metrics).

**Comment by mattnworb** [comment]
Thanks for the explanation. So it would be sufficient for you if Helios set these labels when creating the containers but didn't do anything else with the labels, such as letting you query for jobs that match certain labels, expose this in CLI commands, etc etc ?

**Comment by carlpett** [comment]
Yup, we're pretty much only interested in the creation part, the inspection parts aren't really necessary for us.

File: helios-tools/src/main/java/com/spotify/helios/cli/command/JobCreateCommand.java [hunk 1]
```diff
@@ -541,6 +548,13 @@ int run(final Namespace options, final HeliosClient client, final PrintStream ou
       builder.setDropCapabilities(dropCaps);
     }
 
+    final Map<String, String> labels = Maps.newHashMap();
```
Comment (carlpett) [inline helios-tools/src/main/java/com/spotify/helios/cli/command/JobCreateCommand.java]: Indeed. Fixed!

**Comment by carlpett** [comment]
Hm, did my change actually cause that, or do the tests sometime time out?

**Comment by mattnworb** [comment]
@carlpett no, the tests are a bit flaky - I restarted the CircleCI build and it looks green now.

For the actual contents of the PR, it looks pretty good to me and also like a pretty straightforward change. One aspect of it that I am concerned about is that we have had bad experiences in the past when adding new fields to the Job class that causes the JobId.hash field to change such that an older client/CLI submitting a new Job calculates the hash based on different parameters than the up-to-date server does, which causes the CreateJobResource to respond with an error.

See for instance https://github.com/spotify/helios/pull/1075 and https://github.com/spotify/helios/pull/1076 - but since the JobIdHashIsStableTest test looks like it passes for your change, maybe that is not a huge concern. I'll have to look over the relevant code there again to be sure though.

**Comment by mattnworb** [comment]
One test that would be good to add would be something in TaskConfigTest to verify that the labels are carried through to the ContainerConfig instance as expected. I wouldn't worry too much about running helios-solo or the integration tests (although sorry to hear they don't work well on Windows, although I am not surprised either - we don't have anyone who develops on Windows machines here AFAIK).

**Comment by carlpett** [comment]
Good idea. Looked to me like none of the existing test cases really matched this, so I added a new test that validates that the ContainerConfig matches the job sent to the TaskConfig builder. Currently only checks the new labels field, though.

I think that this change does not break the JobIdHashIsStableTest since it assigns a default empty map, which should be omitted from the json according to the documentation?

**Comment by mattnworb** [comment]
> I think that this change does not break the JobIdHashIsStableTest since it assigns a default empty map, which should be omitted from the json according to the documentation?

Yep that is exactly correct. For future reference (as I'll probably forget some day), [only non-empty fields are used in the `Json.sha1digest(object)` method](https://github.com/spotify/helios/blob/e4374120393cdb8c7f65382ef30a922cc66a074b/helios-client/src/main/java/com/spotify/helios/common/Json.java#L329) that `Job.Builder` calls to compute the hash of a job.

**Comment by mattnworb** [comment]
@carlpett thanks again for the contribution - this should be available in `0.9.123`.

**Pull request merged**